### PR TITLE
Fix all reported clippy warnings

### PR DESCRIPTION
**Pull request by dnaka91**
Currently clippy reports many warnings. This PR fixes all of them with the exception of a  
few being allowed now as it bring any benefit to adjust to them.

While doing so I replaced many usages of `PathBuf` with `&Path` as well.

### CODE REVIEW

File: src/reader.rs [hunk 1]
```diff
@@ -361,12 +361,11 @@ impl Debug for GCNO {
...
 }
 
-impl GCNO {
+impl Gcno {
     pub fn new() -> Self {
-        GCNO {
+        Gcno {
             version: 0,
             checksum: 0,
-            cwd: None,
```
Comment (calixteman) [inline src/reader.rs:369]: I'm not sure that's a good idea to remove this property even if it's unused.

Comment (dnaka91) [reply src/reader.rs:369]: I stumbled over this at first and quickly got some test failures.

Everything is good as long as the value for still read [here](https://github.com/mozilla/grcov/pull/625/files#diff-267c9f4da66412a9f439ac08d224356fe24265b5e1cebb6c44c2d55b96414513R454). I simply removed it from the struct but the reading/parsing logic _should_ not be affected.

Of course I can put it back in if you prefer.

Comment (calixteman) [reply src/reader.rs:369]: Yep I saw that the value is still read.
It'd prefer keeping it in case we need it or even for debug purposes.

Comment (dnaka91) [reply src/reader.rs:369]: I put it back in now. Interestingly clippy didn't warn about it anymore now 🤔

File: src/output.rs [hunk 2]
```diff
@@ -96,7 +96,7 @@ pub fn output_activedata_etl(results: CovResultIter, output_file: Option<&str>,
         for function in result.functions.values() {
             start_indexes.push(function.start);
         }
-        start_indexes.sort();
+        start_indexes.sort_unstable();
```
Comment (calixteman) [inline src/output.rs:99]: I suppose that all `function.start` are different so I guess that using `sort_unstable` doesn't make any difference in the final order with `sort`. Out of curiosity, do you have a link on some benchmarks to compare perf for an array with unique elements ?
Anyway we should maybe use a `btreeset` here.

File: src/parser.rs [hunk 3]
```diff
@@ -529,10 +531,15 @@ fn get_xml_attribute<R: BufRead>(
     )))
 }
 
+struct JacocoReport {
```
Comment (calixteman) [inline src/parser.rs:534]: nit: we use to have this kind of stuff in defs.rs.

File: src/parser.rs [hunk 4]
```diff
@@ -865,11 +873,11 @@ mod tests {
         assert!(result.functions.contains_key("MainProcessSingleton"));
         let func = result.functions.get("MainProcessSingleton").unwrap();
         assert_eq!(func.start, 15);
-        assert_eq!(func.executed, true);
+        assert!(func.executed);
```
Comment (calixteman) [inline src/parser.rs:876]: We'll lose the benefit of using `pretty_assertions` here, but do we care ?? @marco-c what's your opinion ?

**Review by calixteman — APPROVED** [review]
All in all, looks good for me with all tests green.
Thanks a lot for doing that.

File: src/parser.rs (hunk 4 above)
Comment (marco-c) [reply src/parser.rs:876]: I guess they're very few cases, so we can live with it if it's what clippy wants

File: src/output.rs (hunk 2 above)
Comment (dnaka91) [reply src/output.rs:99]: I don't have any benchmarks for that, mostly just followed clippy's recommendations.
It probably doesn't make a difference in this case and just satisfies Google 😅

File: src/parser.rs (hunk 4 above)
Comment (dnaka91) [reply src/parser.rs:876]: Alternatively I can simply turn of that specific clippy lint if that's more preferable.

File: src/parser.rs (hunk 3 above)
Comment (dnaka91) [reply src/parser.rs:534]: I moved it over to defs.rs 🙇

**Comment by dnaka91** [comment]
I have updated the PR to fix the windows build errors and added a few fixes for lints coming from the nightly version of clippy.
Anything else to do for this to be merged @calixteman?

I have a few more changes piled up and based on this PR. Would be lovely if this could be merged soonish 🙇‍♂️

**Comment by marco-c** [comment]
@dnaka91 I think we're good, could you rebase on top of current master?

**Comment by dnaka91** [comment]
@marco-c done, just rebased on top of master 🙇

**Review by marco-c — APPROVED** [review]

**Pull request merged**